### ISSUE TITLE
Should we try integrating Review Ninja?

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcloud-common#31]
Who thinks it'd be a good idea to make master a protected branch and integrate Review Ninja ? (http://review.ninja)

Reply with a vote? Yes/No?

Votes (will be updated by creator / admins based on comments)

Who | for language | Vote
---|---|---
@okdave|Go|No
@broady|Go|No
@aozarov|Java|No
@ajkannan|Java|No
@mziccard|Java|?
@stephenplusplus|Node|?
@callmehiphop|Node|?
@dhermes|Python|No
@tseaver|Python|?
@blowmage|Ruby|No
@quartzmo|Ruby|No

**Comment by broady** [comment]
I don't have a vote, but will weigh in. Looks pretty nice to me, except doesn't appear to provide control over merge strategy.

**Comment by ajkannan** [comment]
I vote no, mostly because I don't think it'd add a lot of functionality for the Java repo, considering how reviews and comments are handled already.

**Comment by okdave** [comment]
+1 to @broady 
We're currently using gerrit for Go, which provides better reviews than PRs and also CLA enforcement etc. However, it's pretty clunky and many devs are not familiar with it.

**Comment by dhermes** [comment]
It looks really pretty, but I don't have much (any?) actual experience with it. For most PRs, GitHub works just fine (flawlessly?) for us.

As far as @broady comment about merge strategy, it'd be nice if we could control that (I prefer "straight line" history, i.e. only allow rebase commits when merging a PR rather than a merge commit from the side).

**Comment by broady** [comment]
@dhermes curious - what is your preferences? squashed merge? or all of the commits rebased AND have a merge commit? or to have a fast-forward merge for single-commit PRs?

**Comment by dhermes** [comment]
I haven't thought about that squashed vs. non-squashed too deeply. I'm inclined to say non-squashed because the commits preserve some history of human behavior. But it requires actively manicuring commits as they go, which can sometime conflict with or slow down the review process.

I was spoiled from the way google3 behaved and never real though about the individual commits as mattering to history, since the review holds all the relevant bits.

As for rebase AND a merge commit, that is a definite nope. Merge commits belong outside the land of "straight line" history.

**Comment by mziccard** [comment]
I have never used it so it's hard to have an opinion. I assume that comments to code made in review.ninja are replicated back in the Github's PR thread, can someone confirm that?

**Comment by jgeewax** [comment]
@broady : You can have a vote ! No worries.

Keep in mind, the goal of Review Ninja is just saying "we explicitly need a :+1: to merge". If you combine RN with protected branches (ie, we say master needs to pass the RN check), then this means "master needs the explicit :+1: ".

The merge strategy is ... basically the same as it is today: if you don't like the strategy, don't give a :+1: ... We can add a script to check that if it's really important, but I don't think it is so far (and will likely change over time).

So to summarize, what I'm hearing so far:

@broady : "This doesn't let me block based on merge strategy" (we don't have the capability today in the regular GH PR flow)

@ajkannan, @dhermes  : "We are fine with what we have today" (makes sense to me, I suspect we'll end up wanting something closer to gerrit, but not gerrit due to the scary amount of overhead)

**Comment by blowmage** [comment]
I'm in the "We are fine with what we have today" camp, which is why I voted "No". We already do this process informally on gcloud-ruby, and I don't see the need for an external tool. Just my opinion, FWIW.

**Comment by aozarov** [comment]
@mziccard Yes, it looks like it.

I played with it a little bit (admittedly not much) as I am very [unhappy](https://github.com/GoogleCloudPlatform/gcloud-java/issues/136) from the github code review tool.

My impression is that this tool *can* help with the merge process, but this part was not a problem so far (though that may change once we get more contributors and owners).

The tool does not seem to provide a better alternative for the code review process itself (actually quite the opposite - the diff is weaker, changes in github took some time to reflect, no per commit diffs,...) but
it does integrate better than other (better in my mind) code review tools (such as [reviewable](https://reviewable.io)).

So, my vote is no.

**Comment by quartzmo** [comment]
I am also "fine with what we have today", and waiting to see what the future brings.

**Comment by jgeewax** [comment]
It's looking like we're heading firmly toward "let's leave this alone for now"-land.  :) Thanks all.
